File: impl/ts/src/estimator.ts
// Hand-written deterministic Path-B estimator arithmetic. Shared constants live in the generated
// estimator_constants.ts; selectivity folds and runtime-unit accounting remain native per core.

import {
  type EstimatorFraction,
  ESTIMATOR_ACCESS_PATH_ORDER,
  ESTIMATOR_UNIT_COUNT,
  ESTIMATOR_UNIT_WEIGHTS,
  MAX_ESTIMATE,
  UNIT_GIN_ENTRY,
  UNIT_GIST_DESCENT,
  UNIT_OPERATOR_EVAL,
  UNIT_PAGE_READ,
  UNIT_ROW_PRODUCED,
  UNIT_STORAGE_ROW_READ,
  SELECTIVITY_BOOLEAN,
  SELECTIVITY_EQUALITY,
  SELECTIVITY_INEQUALITY,
  SELECTIVITY_MATCH,
  SELECTIVITY_MATCHING,
  SELECTIVITY_NULL_TEST,
  SELECTIVITY_OPAQUE,
  SELECTIVITY_PAIRED_RANGE,
} from "./estimator_constants.ts";

export type Selectivity =
  | { kind: "all" | "zero" | "unique" }
  | { kind: "fraction"; fraction: EstimatorFraction }
  | { kind: "not"; child: Selectivity }
  | { kind: "and" | "or"; lhs: Selectivity; rhs: Selectivity };

export function fractionSelectivity(fraction: EstimatorFraction): Selectivity {
  return { kind: "fraction", fraction };
}

export function andSelectivity(lhs: Selectivity, rhs: Selectivity): Selectivity {
  return { kind: "and", lhs, rhs };
}

export function orSelectivity(lhs: Selectivity, rhs: Selectivity): Selectivity {
  return { kind: "or", lhs, rhs };
}

export function notSelectivity(child: Selectivity): Selectivity {
  return { kind: "not", child };
}

export function saturatingEstimateAdd(a: bigint, b: bigint): bigint {
  const sum = a + b;
  return sum > MAX_ESTIMATE ? MAX_ESTIMATE : sum;
}

export function saturatingEstimateMultiply(a: bigint, b: bigint): bigint {
  if (a < 0n || b < 0n || (a !== 0n && b > MAX_ESTIMATE / a)) return MAX_ESTIMATE;
  return a * b;
}

// Exact ceil(a*b/d). BigInt supplies the cross-core wide intermediate directly.
export function ceilEstimateMultiplyDivide(a: bigint, b: bigint, d: bigint): bigint {
  if (a <= 0n || b <= 0n || d <= 0n) return 0n;
  const quotient = (a * b + d - 1n) / d;
  return quotient > MAX_ESTIMATE ? MAX_ESTIMATE : quotient;
}

// ceil(n*numerator/denominator), deliberately expressed without relying on a wider temporary.
export function scaleEstimateCeil(n: bigint, fraction: EstimatorFraction): bigint {
  if (n <= 0n || fraction.numerator <= 0n) return 0n;
  const quotient = n / fraction.denominator;
  const remainder = n % fraction.denominator;
  const whole = saturatingEstimateMultiply(quotient, fraction.numerator);
  const product = saturatingEstimateMultiply(remainder, fraction.numerator);
  const tail = product / fraction.denominator + (product % fraction.denominator === 0n ? 0n : 1n);
  return saturatingEstimateAdd(whole, tail);
}

function clampEstimate(value: bigint): bigint {
  if (value < 0n) return 0n;
  return value > MAX_ESTIMATE ? MAX_ESTIMATE : value;
}

export function estimateSelectivity(selectivity: Selectivity, inputRows: bigint): bigint {
  const n = clampEstimate(inputRows);
  switch (selectivity.kind) {
    case "all":
      return n;
    case "zero":
      return 0n;
    case "unique":
      return n > 0n ? 1n : 0n;
    case "fraction":
      return scaleEstimateCeil(n, selectivity.fraction);
    case "not":
      return n - estimateSelectivity(selectivity.child, n);
    case "and":
      return estimateSelectivity(selectivity.rhs, estimateSelectivity(selectivity.lhs, n));
    case "or":
      return saturatingEstimateAdd(
        estimateSelectivity(selectivity.lhs, n),
        estimateSelectivity(selectivity.rhs, n),
      ) > n
        ? n
        : saturatingEstimateAdd(
            estimateSelectivity(selectivity.lhs, n),
            estimateSelectivity(selectivity.rhs, n),
          );
  }
}

export function selectivityClass(classification: string): Selectivity {
  switch (classification) {
    case "equality":
      return fractionSelectivity(SELECTIVITY_EQUALITY);
    case "inequality":
      return fractionSelectivity(SELECTIVITY_INEQUALITY);
    case "paired_range":
      return fractionSelectivity(SELECTIVITY_PAIRED_RANGE);
    case "null_test":
      return fractionSelectivity(SELECTIVITY_NULL_TEST);
    case "match":
      return fractionSelectivity(SELECTIVITY_MATCH);
    case "matching":
      return fractionSelectivity(SELECTIVITY_MATCHING);
    case "boolean":
      return fractionSelectivity(SELECTIVITY_BOOLEAN);
    default:
      return fractionSelectivity(SELECTIVITY_OPAQUE);
  }
}

export type CandidateEstimate = {
  rows: bigint;
  units: bigint[];
  cost: bigint;
  tieKey: string;
};

export type PlanEstimate = {
  rows: bigint;
  logicalRows: bigint;
  units: bigint[];
};

export type EstimatedPlan = { root: PlanEstimate; nodes: PlanEstimate[] };

export function emptyPlanEstimate(rows = 0n, logicalRows = rows): PlanEstimate {
  return {
    rows: clampEstimate(rows),
    logicalRows: clampEstimate(logicalRows),
    units: Array<bigint>(ESTIMATOR_UNIT_COUNT).fill(0n),
  };
}

export function clonePlanEstimate(estimate: PlanEstimate): PlanEstimate {
  return { rows: estimate.rows, logicalRows: estimate.logicalRows, units: [...estimate.units] };
}

export function planEstimateCost(estimate: PlanEstimate): bigint {
  let cost = 0n;
  for (let i = 0; i < ESTIMATOR_UNIT_COUNT; i++) {
    cost = saturatingEstimateAdd(
      cost,
      saturatingEstimateMultiply(estimate.units[i]!, ESTIMATOR_UNIT_WEIGHTS[i]!),
    );
  }
  return cost;
}

export function addPlanEstimates(lhs: PlanEstimate, rhs: PlanEstimate): PlanEstimate {
  const out = clonePlanEstimate(lhs);
  for (let i = 0; i < ESTIMATOR_UNIT_COUNT; i++)
    out.units[i] = saturatingEstimateAdd(out.units[i]!, rhs.units[i]!);
  return out;
}

export function repeatPlanEstimate(estimate: PlanEstimate, count: bigint): PlanEstimate {
  const n = clampEstimate(count);
  return {
    rows: saturatingEstimateMultiply(estimate.rows, n),
    logicalRows: saturatingEstimateMultiply(estimate.logicalRows, n),
    units: estimate.units.map((value) => saturatingEstimateMultiply(value, n)),
  };
}

export function addPlanUnit(estimate: PlanEstimate, unit: number, count: bigint): void {
  estimate.units[unit] = saturatingEstimateAdd(estimate.units[unit]!, clampEstimate(count));
}

export function leafEstimatedPlan(estimate: PlanEstimate): EstimatedPlan {
  return { root: estimate, nodes: [estimate] };
}

export function wrapEstimatedPlan(
  child: EstimatedPlan,
  rows: bigint,
  logicalRows: bigint,
  local: bigint[],
): EstimatedPlan {
  const root = clonePlanEstimate(child.root);
  root.rows = clampEstimate(rows);
  root.logicalRows = clampEstimate(logicalRows);
  for (let i = 0; i < ESTIMATOR_UNIT_COUNT; i++)
    root.units[i] = saturatingEstimateAdd(root.units[i]!, local[i] ?? 0n);
  return { root, nodes: [root, ...child.nodes] };
}

export function parentEstimatedPlan(
  root: PlanEstimate,
  ...children: EstimatedPlan[]
): EstimatedPlan {
  const parent = clonePlanEstimate(root);
  return { root: parent, nodes: [parent, ...children.flatMap((child) => child.nodes)] };
}

export type CandidateEstimateInputs = {
  kind: string;
  indexName: string;
  scanRows: bigint;
  outputRows: bigint;
  accessPages: bigint;
  tableHeight: bigint;
  filterNodes: bigint;
  accessWork: bigint;
  producesRows: boolean;
};

export function candidateTieKey(kind: string, indexName: string): string {
  const found = (ESTIMATOR_ACCESS_PATH_ORDER as readonly string[]).indexOf(kind);
  const rank = found < 0 ? ESTIMATOR_ACCESS_PATH_ORDER.length : found;
  return `${rank}:${indexName}`;
}

export function estimateCandidate(input: CandidateEstimateInputs): CandidateEstimate {
  const scanRows = clampEstimate(input.scanRows);
  const outputRows = clampEstimate(input.outputRows);
  const units = Array<bigint>(ESTIMATOR_UNIT_COUNT).fill(0n);
  units[UNIT_STORAGE_ROW_READ] = scanRows;
  units[UNIT_PAGE_READ] = clampEstimate(input.accessPages);
  if (["btree", "gist", "gin", "index_interval"].includes(input.kind)) {
    units[UNIT_PAGE_READ] = saturatingEstimateAdd(
      units[UNIT_PAGE_READ]!,
      saturatingEstimateMultiply(scanRows, clampEstimate(input.tableHeight)),
    );
  }
  units[UNIT_OPERATOR_EVAL] = saturatingEstimateMultiply(
    scanRows,
    clampEstimate(input.filterNodes),
  );
  if (input.producesRows) units[UNIT_ROW_PRODUCED] = outputRows;
  if (input.kind === "gin") units[UNIT_GIN_ENTRY] = clampEstimate(input.accessWork);
  if (input.kind === "gist") units[UNIT_GIST_DESCENT] = clampEstimate(input.accessWork);
  let cost = 0n;
  for (let i = 0; i < ESTIMATOR_UNIT_COUNT; i++) {
    cost = saturatingEstimateAdd(
      cost,
      saturatingEstimateMultiply(units[i]!, ESTIMATOR_UNIT_WEIGHTS[i]!),
    );
  }
  return { rows: outputRows, units, cost, tieKey: candidateTieKey(input.kind, input.indexName) };
}

File: impl/rust/src/estimator.rs
//! Deterministic Path-B estimator arithmetic shared by planner shadow estimates and fixtures.

use crate::estimator_constants::*;

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum Selectivity {
    All,
    Zero,
    Unique,
    Fraction(EstimatorFraction),
    Not(Box<Selectivity>),
    And(Box<Selectivity>, Box<Selectivity>),
    Or(Box<Selectivity>, Box<Selectivity>),
}

impl Selectivity {
    pub(crate) fn fraction(fraction: EstimatorFraction) -> Self {
        Self::Fraction(fraction)
    }

    pub(crate) fn and(self, rhs: Self) -> Self {
        Self::And(Box::new(self), Box::new(rhs))
    }

    pub(crate) fn or(self, rhs: Self) -> Self {
        Self::Or(Box::new(self), Box::new(rhs))
    }

    pub(crate) fn not(self) -> Self {
        Self::Not(Box::new(self))
    }
}

pub(crate) fn sat_add(a: i64, b: i64) -> i64 {
    a.saturating_add(b).min(MAX_ESTIMATE)
}

pub(crate) fn sat_mul(a: i64, b: i64) -> i64 {
    a.saturating_mul(b).min(MAX_ESTIMATE)
}

/// `ceil(n * numerator / denominator)` without a wider intermediate.
pub(crate) fn scale_ceil(n: i64, fraction: EstimatorFraction) -> i64 {
    debug_assert!(n >= 0 && fraction.numerator >= 0 && fraction.denominator > 0);
    if n == 0 || fraction.numerator == 0 {
        return 0;
    }
    let quotient = n / fraction.denominator;
    let remainder = n % fraction.denominator;
    let whole = sat_mul(quotient, fraction.numerator);
    let tail_product = sat_mul(remainder, fraction.numerator);
    let tail =
        tail_product / fraction.denominator + i64::from(tail_product % fraction.denominator != 0);
    sat_add(whole, tail)
}

pub(crate) fn estimate_rows(selectivity: &Selectivity, input_rows: i64) -> i64 {
    let n = input_rows.clamp(0, MAX_ESTIMATE);
    match selectivity {
        Selectivity::All => n,
        Selectivity::Zero => 0,
        Selectivity::Unique => n.min(1),
        Selectivity::Fraction(fraction) => scale_ceil(n, *fraction),
        Selectivity::Not(child) => n - estimate_rows(child, n),
        Selectivity::And(lhs, rhs) => {
            let left_rows = estimate_rows(lhs, n);
            estimate_rows(rhs, left_rows)
        }
        Selectivity::Or(lhs, rhs) => sat_add(estimate_rows(lhs, n), estimate_rows(rhs, n)).min(n),
    }
}

pub(crate) fn selectivity_class(class: &str) -> Selectivity {
    match class {
        "equality" => Selectivity::fraction(SELECTIVITY_EQUALITY),
        "inequality" => Selectivity::fraction(SELECTIVITY_INEQUALITY),
        "paired_range" => Selectivity::fraction(SELECTIVITY_PAIRED_RANGE),
        "null_test" => Selectivity::fraction(SELECTIVITY_NULL_TEST),
        "match" => Selectivity::fraction(SELECTIVITY_MATCH),
        "matching" => Selectivity::fraction(SELECTIVITY_MATCHING),
        "boolean" => Selectivity::fraction(SELECTIVITY_BOOLEAN),
        _ => Selectivity::fraction(SELECTIVITY_OPAQUE),
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct CandidateEstimate {
    pub(crate) rows: i64,
    pub(crate) units: [i64; ESTIMATOR_UNIT_COUNT],
    pub(crate) cost: i64,
    pub(crate) tie_key: String,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) struct CandidateInputs<'a> {
    pub(crate) kind: &'a str,
    pub(crate) index_name: &'a str,
    pub(crate) scan_rows: i64,
    pub(crate) output_rows: i64,
    pub(crate) access_pages: i64,
    pub(crate) table_height: i64,
    pub(crate) filter_nodes: i64,
    pub(crate) access_work: i64,
    pub(crate) produces_rows: bool,
}

fn access_rank(kind: &str) -> usize {
    ACCESS_PATH_ORDER
        .iter()
        .position(|candidate| *candidate == kind)
        .unwrap_or(ACCESS_PATH_ORDER.len())
}

pub(crate) fn candidate_tie_key(kind: &str, index_name: &str) -> String {
    format!("{}:{}", access_rank(kind), index_name)
}

pub(crate) fn estimate_candidate(input: CandidateInputs<'_>) -> CandidateEstimate {
    let scan_rows = input.scan_rows.clamp(0, MAX_ESTIMATE);
    let output_rows = input.output_rows.clamp(0, MAX_ESTIMATE);
    let mut units = [0; ESTIMATOR_UNIT_COUNT];
    units[UNIT_STORAGE_ROW_READ] = scan_rows;
    units[UNIT_PAGE_READ] = input.access_pages.clamp(0, MAX_ESTIMATE);
    if matches!(input.kind, "btree" | "gist" | "gin" | "index_interval") {
        units[UNIT_PAGE_READ] = sat_add(
            units[UNIT_PAGE_READ],
            sat_mul(scan_rows, input.table_height.clamp(0, MAX_ESTIMATE)),
        );
    }
    units[UNIT_OPERATOR_EVAL] = sat_mul(scan_rows, input.filter_nodes.clamp(0, MAX_ESTIMATE));
    if input.produces_rows {
        units[UNIT_ROW_PRODUCED] = output_rows;
    }
    if input.kind == "gin" {
        units[UNIT_GIN_ENTRY] = input.access_work.clamp(0, MAX_ESTIMATE);
    }
    if input.kind == "gist" {
        units[UNIT_GIST_DESCENT] = input.access_work.clamp(0, MAX_ESTIMATE);
    }
    let cost = units
        .iter()
        .zip(ESTIMATOR_UNIT_WEIGHTS)
        .fold(0, |total, (count, weight)| {
            sat_add(total, sat_mul(*count, weight))
        });
    CandidateEstimate {
        rows: output_rows,
        units,
        cost,
        tie_key: candidate_tie_key(input.kind, input.index_name),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn named_selectivity(token: &str) -> Selectivity {
        match token {
            "all" => Selectivity::All,
            "zero" => Selectivity::Zero,
            "unique" => Selectivity::Unique,
            "equality" => Selectivity::fraction(SELECTIVITY_EQUALITY),
            "inequality" => Selectivity::fraction(SELECTIVITY_INEQUALITY),
            "paired_range" => Selectivity::fraction(SELECTIVITY_PAIRED_RANGE),
            "null_test" => Selectivity::fraction(SELECTIVITY_NULL_TEST),
            "match" => Selectivity::fraction(SELECTIVITY_MATCH),
            "matching" => Selectivity::fraction(SELECTIVITY_MATCHING),
            "boolean" => Selectivity::fraction(SELECTIVITY_BOOLEAN),
            "opaque" => Selectivity::fraction(SELECTIVITY_OPAQUE),
            _ => panic!("unknown selectivity token {token}"),
        }
    }

    fn postfix(tokens: &[toml::Value]) -> Selectivity {
        let mut stack = Vec::new();
        for value in tokens {
            let token = value.as_str().expect("token string");
            match token {
                "not" => {
                    let child = stack.pop().expect("not operand");
                    stack.push(Selectivity::not(child));
                }
                "and" | "or" => {
                    let rhs = stack.pop().expect("right operand");
                    let lhs = stack.pop().expect("left operand");
                    stack.push(if token == "and" {
                        lhs.and(rhs)
                    } else {
                        lhs.or(rhs)
                    });
                }
                _ => stack.push(named_selectivity(token)),
            }
        }
        assert_eq!(stack.len(), 1);
        stack.pop().unwrap()
    }

    #[test]
    fn shared_estimator_vectors() {
        let path = concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/../../spec/cost/estimator_vectors.toml"
        );
        let source = std::fs::read_to_string(path).unwrap();
        let root: toml::Value = toml::from_str(&source).unwrap();
        for row in root["arithmetic"].as_array().unwrap() {
            let a = row["a"].as_integer().unwrap();
            let b = row["b"].as_integer().unwrap();
            let actual = match row["op"].as_str().unwrap() {
                "sat_add" => sat_add(a, b),
                "sat_mul" => sat_mul(a, b),
                "scale_ceil" => scale_ceil(
                    a,
                    EstimatorFraction {
                        numerator: b,
                        denominator: row["c"].as_integer().unwrap(),
                    },
                ),
                op => panic!("unknown arithmetic op {op}"),
            };
            assert_eq!(
                actual,
                row["expected"].as_integer().unwrap(),
                "{}",
                row["id"]
            );
        }
        for row in root["predicate"].as_array().unwrap() {
            let expr = postfix(row["tokens"].as_array().unwrap());
            assert_eq!(
                estimate_rows(&expr, row["n"].as_integer().unwrap()),
                row["expected"].as_integer().unwrap(),
                "{}",
                row["id"]
            );
        }
        for row in root["candidate"].as_array().unwrap() {
            let estimate = estimate_candidate(CandidateInputs {
                kind: row["kind"].as_str().unwrap(),
                index_name: row["index_name"].as_str().unwrap(),
                scan_rows: row["scan_rows"].as_integer().unwrap(),
                output_rows: row["output_rows"].as_integer().unwrap(),
                access_pages: row["access_pages"].as_integer().unwrap(),
                table_height: row["table_height"].as_integer().unwrap(),
                filter_nodes: row["filter_nodes"].as_integer().unwrap(),
                access_work: row["access_work"].as_integer().unwrap(),
                produces_rows: row["produces_rows"].as_bool().unwrap(),
            });
            assert_eq!(
                estimate.rows,
                row["est_rows"].as_integer().unwrap(),
                "{}",
                row["id"]
            );
            assert_eq!(
                estimate.cost,
                row["est_cost"].as_integer().unwrap(),
                "{}",
                row["id"]
            );
            assert_eq!(
                estimate.tie_key,
                row["tie_key"].as_str().unwrap(),
                "{}",
                row["id"]
            );
            let expected_units = row["units"].as_table().unwrap();
            for (i, id) in ESTIMATOR_UNIT_IDS.iter().enumerate() {
                let expected = expected_units
                    .get(*id)
                    .and_then(toml::Value::as_integer)
                    .unwrap_or(0);
                assert_eq!(estimate.units[i], expected, "{} unit {id}", row["id"]);
            }
        }
    }
}
